src/main/resources/data/computercraft/lua/rom/modules/main/cc/shell/completion.lua

```diff
@@ -74,12 +76,41 @@ end
 
 --- Complete the name of a program.
 --
--- @tparam table shell The shell we're completing in
--- @tparam { string... } choices The list of choices to complete from.
+-- @tparam table shell The shell we're completing in.
+-- @tparam string text Current text to complete.
+-- @tparam { string... } previous The shell arguments before this one.
+-- @tparam[opt] boolean completion_space Whether to add a space after the completed item if there is completion for this program.
+-- @treturn { string... } A list of suffixes of matching programs.
+-- @see shell.completeProgram
+local function program(shell, text, previous, completion_space)
+    local results = shell.completeProgram(text)
+    if completion_space then
+        local tCompletionInfo = shell.getCompletionInfo()
+        for n = 1, #results do
+            local sResult = results[n]
+            if sResult:sub(-1) ~= "/" and tCompletionInfo[shell.resolveProgram(text .. sResult)] then
+                results[n] = sResult .. " "
+            end
+        end
+    end
+    return results
+end
+
+--- Complete arguments of a program.
+--
+-- @tparam table shell The shell we're completing in.
+-- @tparam string text Current text to complete.
+-- @tparam { string... } previous The shell arguments before this one.
+-- @tparam string|number program Path of program to complete args for, or a number of argument index which specified program.
+-- @tparam number starting Which argument index this program args started at.
 -- @treturn { string... } A list of suffixes of matching programs.
 -- @see shell.completeProgram
-local function program(shell, text)
-    return shell.completeProgram(text)
+local function programArgs(shell, text, previous, program, starting)
+    local resolved = type(program) == "string" and program or shell.resolveProgram(previous[program])
+    if not resolved then return end
+    local tCompletionInfo = shell.getCompletionInfo()
```

Comment:
Nitpick, but `shell.getCompletionInfo()[resolved]` instead? Saves the additional lookup.